File: comphysics_julia/trapezoidal.jl
################################################################################
# Title:         Trapezoidal Rule Integration in Julia
# Description:   This program approximates the definite integral of a given
#                function f(x) over the interval [a, b] using the trapezoidal
#                rule. The function 'trapezoidal' computes and returns the
#                approximated value given the number of subintervals n.
#
# Instructions:
#   1. Save this file as, for example, trapezoidal.jl.
#   2. To run the program, open a terminal and navigate to the file's directory.
#   3. Execute the program with the command: julia trapezoidal.jl
################################################################################

# The trapezoidal rule to approximate the definite integral of function f
#          over the interval [a, b] using n subdivisions.
function trapezoidal(a, b, n, f)
    # Calculate the width of each subinterval.
    h = (b - a) / n

    # Initialize sum 's' with the average of the function values at the endpoints.
    s = (f(a) + f(b)) / 2

    # Loop over each interior point (i=1 to n-1) and accumulate the function values.
    for i in 1:(n - 1)
        s += f(a + i * h)  # Evaluate f at the current x value and add to the sum.
    end

    # Multiply the accumulated sum with the width 'h' to get the final approximation.
    y = h * s

    # Return the approximated integral value.
    return y
end

# Define the function 'f(x)' as an anonymous function.
# In this example, f(x) = x^2 * sin(x),
f = x -> x^2 * sin(x);

# If you wish to test a different function, simply modify the definition above.
# For example, you can try:
# f = t -> t^2 + 1 or anything

# Compute the approximate integral of f over the interval [0, 1] using 4 subintervals.
y = trapezoidal(0, 1, 4, f)

# Print the result to the console.
println("Trapezoidal; y= ", y)

File: comphysics_octave/trapezoidal.m
% Function to approximate the integral of a given function f using the trapezoidal rule.
% Inputs:
%   a - Left endpoint of the interval.
%   b - Right endpoint of the interval.
%   n - Number of subintervals in which the interval [a,b] is divided.
%   f - Function handle representing the function to integrate.
% Output:
%   y - Approximated value of the definite integral.

function y = trapezoidal(a, b, n, f)
  % Compute the width (h) of each subinterval.
  h = (b-a)/n;

  % Initialize the sum (s) with the average of the function values at the endpoints.
  s = ((f(a) + f(b))/2);

  % Loop over the interior points of the interval.
  for i = 1:(n-1)
    % Add the function value at each interior point to the sum.
    % a + i*h computes the current x-value.
    s += f(a + i*h);
  endfor

  % Multiply the sum by the subinterval width to obtain the integral approximation.
  y = h*s;
end

% Compiler/ Interpreter Instruction

% 1. Save the code in a file named, for example, "trapezoidal.m".
% 2. Run GNU Octave (or Octave CLI/GUI).
% 3. In Octave’s command prompt, call the function with the appropriate parameters,
% for example:
% octave:1> result = trapezoidal(0, 1, 100, @(x) x.^2);
